ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/LogSegment.java

```diff
@@ -281,12 +281,16 @@ public ReferenceCountedObject<LogEntryProto> load(LogRecord key) throws IOExcept
       final AtomicReference<ReferenceCountedObject<LogEntryProto>> toReturn = new AtomicReference<>();
       final LogSegmentStartEnd startEnd = LogSegmentStartEnd.valueOf(startIndex, endIndex, isOpen);
       readSegmentFile(file, startEnd, maxOpSize, getLogCorruptionPolicy(), raftLogMetrics, entryRef -> {
-        final LogEntryProto entry = entryRef.retain();
-        final TermIndex ti = TermIndex.valueOf(entry);
-        putEntryCache(ti, entryRef, Op.LOAD_SEGMENT_FILE);
-        if (ti.equals(key.getTermIndex())) {
-          toReturn.set(entryRef);
-        } else {
+        try {
+          final LogEntryProto entry = entryRef.retain();
```

Comment:
`entryRef.retain()` should be called before the try-block.  Otherwise, if it throws an exception, we will call `release()` without successfully retained.